weasyl/two_factor_auth.py

```diff
@@ -0,0 +1,264 @@
+    if totp.verify(tfa_response):
+        return tfa_secret, generate_recovery_codes(userid)
+    else:
+        return False
+
+
+def activate(userid, tfa_secret, tfa_response):
+    """
+    Fully activate 2FA for a given user account, after final validation of the TOTP secret.
+
+    This function is part two--the final part--in enabling 2FA. Passing this step ensures that
+    the user has been presented the opportunity to save recovery keys for their account.
+
+    Parameters:
+        userid: The userid of the calling user.
+        tfa_secret: The 2FA secret generated from tfa_init(); retrieved from the
+        verification page's form information.
+        tfa_response: The 2FA challenge-response code to verify against tfa_secret.
+
+    Returns: Boolean True if the `tfa_response` corresponds with `tfa_secret`, thus enabling 2FA,
+        otherwise Boolean False indicating 2FA has not been enabled.
+    """
+    totp = pyotp.TOTP(tfa_secret)
+    # If the provided `tfa_response` matches the TOTP value, write the 2FA secret into `login`, activating 2FA for `userid`
+    if totp.verify(tfa_response):
+        d.engine.execute("""
+            UPDATE login
+            SET twofa_secret = (%(tfa_secret)s)
+            WHERE userid = (%(userid)s)
+        """, tfa_secret=tfa_secret, userid=userid)
+        return True
+    else:
+        return False
+
+
+def verify(userid, tfa_response):
+    """
+    Verify a 2FA-enabled user's 2FA challenge-response against the stored
+    2FA secret.
+
+    Parameters:
+        userid: The userid to compare the 2FA challenge-response against.
+        tfa_response: User-supplied response. May be either the Google Authenticator
+        (or other app) supplied code, or a recovery code
+
+    Returns: Boolean True if 2FA verification is successful, Boolean False otherwise.
+    """
+    tfa_secret = d.engine.scalar("""
+        SELECT twofa_secret
+        FROM login
+        WHERE userid = (%(userid)s)
+    """, userid=userid)
+    # Validate supplied 2FA response versus calculated current TOTP value.
+    totp = pyotp.TOTP(tfa_secret)
+    if totp.verify(tfa_response):
+        return True
+    # TOTP verification failed, check recovery code
+    elif is_recovery_code_valid(userid, tfa_response):
+        # Recovery code was valid, and consumed
+        return True
+    else:
+        # Received input was not a valid TOTP response, and was not a valid recovery code
+        return False
+
+
+def get_number_of_recovery_codes(userid):
+    """
+    Get and return the number of remaining recovery codes for `userid`.
+    
+    Parameters:
+        userid: The userid for which to check the count of recovery codes.
+
+    Returns:
+        An integer representing the number of remaining recovery codes.
+    """
+    return d.engine.scalar("""
+        SELECT COUNT(*)
+        FROM twofa_recovery_codes
+        WHERE userid = (%(userid)s)
+    """, userid=userid)
+
+
+def generate_recovery_codes(userid):
```

Comment:
The font in use was determined in this author's opinion to not be distinct enough. Codes as made in ``generate_recovery_codes()`` are created upper case, (thus being displayed to the user as upper case), and in ``is_recovery_code_valid()`` the received input is force-converted to uppercase via ``.upper()`` to ensure a user entering lower-case codes still validates. Effective security provided, therefore, should be ``20^36`` (digits+numerals; 6.87e+46~) when recovery codes are factored into the equation, compounded by the fact that the user's password is still required for authentication. (AKA, it is not real likely to guess these, even with the entropy reduced by eliminating lower-case characters, or so it appears).